### PR TITLE
Introduce market models & register action

### PR DESCRIPTION
**Pull request by ipdae**
### MarketState
It has the avatar addresses who registered product in the market. It will used when searching for all `ProductsState`.(corresponding to `ShardedShopState`)
### ProductsState
Each AvatarState will have it own a list of registered ProductIds.(corresponding to `OrderDigestListState`)

## Products
registered item, FungibleAssetValue information. (corresponding to `Order, ShopItem`)
### RegisterProduct
register item or FungibleAssetValue to market. (corresponding to `Sell`)
### RegisterInfo
It is used for registering item to the market. 
### AssetInfo
It is used for registering FungibleAssetValue to the market.
### ProductInfo
It is used for find registered product in the market.

## TL;DR
- The difference with the existing Order model is that instead of copying the item, it completely moves the item (or FungibleAssetValue) to product address.
- IRegisterInfo(RegisterInfo, AssetInfo) & ProductInfo using it action plain values. we can register, cancel, buy, re-register by combining the two models.
- This pr is the divided from #1672 into stages for review convenience. If you need information about the entire action and model, you can check it.

### CODE REVIEW

**Review by Unengine — APPROVED** [review]

**Review by hskim881028 — CHANGES_REQUESTED** [review]

File: Lib9c/Action/RegisterProduct.cs [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+using Nekoyume.Model.Market;
+using Nekoyume.Model.State;
+using static Lib9c.SerializeKeys;
+
+namespace Nekoyume.Action
+{
+    [ActionType("register_product")]
+    public class RegisterProduct : GameAction
+    {
+        public IEnumerable<IRegisterInfo> RegisterInfos;
```
[the remark this answers is not in the document]
Comment (ipdae) [reply Lib9c/Action/RegisterProduct.cs:20]: i think it can be check avatar address distinct count check https://github.com/planetarium/lib9c/pull/1733/files#diff-7dfbd7fe2b73e121e06d59b4c51a6b35ad6df442dd5729ef0fbfb08148c28494R30-R33

**Review by hskim881028 — APPROVED** [review]
